install.packages("SNPfiltR")
install.packages("vcfR")
library(SNPfiltR)
library(vcfR)
#read in vcf as vcfR
vcfR <- read.vcfR("~/Research_Smonticola/NAmSorex75.snps.vcf")

### check the metadata present in your vcf
vcfR

#generate popmap file. Two column popmap with the same format as stacks, and the columns must be named 'id' and 'pop'
### DID NOT WORK. Imported popmap file.
popmap<-data.frame(id=colnames(vcfR@gt)[2:length(colnames(vcfR@gt))],pop=substr(colnames(vcfR@gt)[2:length(colnames(vcfR@gt))], 3,11))

### check the metadata present in your vcf
vcfR

popmap <- popmap_NAmSorexOnly

#visualize distributions
hard_filter(vcfR=vcfR)

#execute allele balance filter
vcfR<-filter_allele_balance(vcfR)

#visualize and pick appropriate max depth cutoff
max_depth(vcfR)

#filter vcf by the max depth cutoff you chose
### Depth is way lower than in the tutorial (5.9 mean vs. 46.7)... should this affect the cutoff? 
vcfR<-max_depth(vcfR, maxdepth = 100)
### Only 0.24% of SNPs were filtered w cutoff of 100 (0.15% from n. am. sorex)

#remove invariant SNPs generated during the genotype filtering steps
vcfR<-min_mac(vcfR, min.mac = 1)
### 2.04% filtered w/ min. allele count below 1
### 2.83% with nam sorex

#check vcfR to see how many SNPs we have left
vcfR
###2,538 in n. am. sorex
#run function to visualize samples
missing_by_sample(vcfR=vcfR, popmap = popmap)
###Groups with fewer than two data points have been dropped.

#run function to drop samples above the threshold we want from the vcf
vcfR<-missing_by_sample(vcfR=vcfR, cutoff = .81)
### 13 samples removed at .81
### 19 removed at .61 cutoff (all non-Sorex shrews removed at this level)
### 4 samples removed at .81 from n am sorex
#subset popmap to only include retained individuals
popmap<-popmap[popmap$id %in% colnames(vcfR@gt),]

#remove invariant sites generated by dropping individuals
vcfR<-min_mac(vcfR, min.mac = 1)
# 2.69% filtered (below minor allele count of 1) at .81 cutoff for missing data per sample
# 8.3% removed at .61 cutoff for missing data per sample
# 0.04% from n am sorex at 0.81
#verify that missing data is not driving clustering patterns among the retained samples
miss<-assess_missing_data_pca(vcfR=vcfR, popmap = popmap, thresholds = .8, clustering = FALSE)
###22.45% of SNPs fell below a completeness cutoff of 0.8 and were removed from the VCF
### 2.67% of SNPs fell below a completeness cutoff of 0.8 and were removed from the VCF

#visualize missing data by SNP and the effect of various cutoffs on the missingness of each sample
missing_by_snp(vcfR)
missing_by_snp <- missing_by_snp(vcfR)

#check if these cutoffs prevent samples from clustering
#by patterns of missing data in a PCA
miss<-assess_missing_data_pca(vcfR=vcfR, popmap = popmap,
                              thresholds = c(.75,.85,.95), clustering = FALSE)

#investigate clustering patterns with and without a minor allele cutoff
#use min.mac() to investigate the effect of multiple cutoffs
vcfR.mac<-min_mac(vcfR = vcfR, min.mac = 2)


#assess clustering without MAC cutoff
miss<-assess_missing_data_tsne(vcfR, popmap, clustering = FALSE)

#assess clustering with MAC cutoff
miss<-assess_missing_data_tsne(vcfR.mac, popmap, clustering = FALSE)

#plot depth per snp and per sample
dp <- extract.gt(vcfR, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)

#plot genotype quality per snp and per sample
gq <- extract.gt(vcfR, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)

#write out vcf with all SNPs
vcfR::write.vcf(vcfR, "~/Research_Smonticola/NAmSorex.filtered.vcf.gz")

#linkage filter vcf to thin SNPs to one per 500bp
vcfR.thin<-distance_thin(vcfR, min.distance = 500)

#write out thinned vcf
vcfR::write.vcf(vcfR.thin, "~/Research_Smonticola/NAmSorex.filtered.thinned.vcf.gz")

write.table(popmap,  file = "popmap_NAmSorex.txt", quote = FALSE, sep ="\t" ,row.names = TRUE, col.names = TRUE)
