####READ ME####
# The purpose of this script is to run SNPfiltR protocol
####libraries####
library(SNPfiltR)
library(devtools)
devtools::install_github("DevonDeRaad/SNPfiltR")
library(vcfR)
####Read in files####
#VCF file
vcfR <- read.vcfR("rbrush/vcf/populations4.snps.vcf")
#populations map file
popmap<-data.frame(id=colnames(vcfR@gt)[2:length(colnames(vcfR@gt))],pop=substr(colnames(vcfR@gt)[2:length(colnames(vcfR@gt))], 1,8))
####Implement quality filters####
#that donâ€™t involve missing data This is because removing low data samples will alter percentage/quantile based missing data cutoffs, so we wait to implement those until after deciding on our final set of samples for downstream analysis
#visualize distributions
hard_filter(vcfR=vcfR)
#> no depth cutoff provided, exploratory visualization will be generated.
#1. genotype depyh: most of my depth (covereage) is below 200. all my data is by 20X
#2. leses 25 genotypes are by 20X
#3. froom the phred score. this is my genotype quality. we set a quality phred of pradtags
#hard filter to minimum depth of 5, and minimum genotype quality of 30
vcfR<-hard_filter(vcfR=vcfR, depth = 5, gq = 30)
#>0.01% of genotypes fall below a read depth of 5 and were converted to NA
#>0.21% of genotypes fall below a genotype quality of 30 and were converted to NA
##execute allele balance filter
#has to do with heterozygotes. if theere are too many and they are in non equilibrium.
#I fon't want them in extreemes, only in 10% or 90% of my samples.
#I need to understand if I want them in my data set. if I want to look at selection, I maybe do want tem, if i only want pop structure, myabe not.
vcfR<-filter_allele_balance(vcfR)
#>20.38% of het genotypes (0.88% of all genotypes) fall outside of 0.25 - 0.75 allele balance ratio and were converted to NA
#I hve some verey high heterozygosity. Somethign weird
#if I want to look at them looko just at loci between 75 - 90
#keep the middle loci if I want to look at population structure
#visualize and pick appropriate max depth cutoff
max_depth(vcfR)
#this is the coveerage for SNPS. most of the data set is above 10 20X
#filter vcf by the max depth cutoff you chose
vcfR<-max_depth(vcfR, maxdepth = 100)
#look at max depth before cutting here. for this cut to at least 100, play with it and see how it influences downstream analysis
#remove invariant SNPs generated during the genotype filtering steps
vcfR<-min_mac(vcfR, min.mac = 1)
#>8.68% of SNPs fell below a minor allele count of 1 and were removed from the VCF
#check vcfR to see how many SNPs we have left
vcfR
####cutoff for missing data allowed per sample####
#run function to visualize samples
missing_by_sample(vcfR=vcfR, popmap = popmap)
#1.
#2.this helps me decide where I want my threshold to be. if i go over 80, I'll loose a bunch of populations
#we want to keep enough samples to measure Fst for example
#3.proportion ofmissing data per individual
#run function to drop samples above the threshold we want from the vcf
#vcfR<-missing_by_sample(vcfR=vcfR, cutoff = .6)
#i wouldn't cut off. 60% missing data is not too bad since O's are outgroup
#>3 samples are above a 0.6 missing data cutoff, and were removed from VCF
#subset popmap to only include retained individuals
popmap<-popmap[popmap$id %in% colnames(vcfR@gt),]
#remove invariant sites generated by dropping individuals
vcfR<-min_mac(vcfR, min.mac = 1)
#>10.16% of SNPs fell below a minor allele count of 1 and were removed from the VCF
#verify that missing data is not driving clustering patterns among the retained samples
miss<-assess_missing_data_pca(vcfR=vcfR, popmap = popmap, thresholds = .8, clustering = FALSE)
#>5.04% of SNPs fell below a completeness cutoff of 0.8 and were removed from the VCF
####set missing data per SNP cutoff####
#visualize missing data by SNP and the effect of various cutoffs on the missingness of each sample
missing_by_snp(vcfR)
#>Picking joint bandwidth of 0.0484
#check if these cutoffs prevent samples from clustering
#by patterns of missing data in a PCA
miss<-assess_missing_data_pca(vcfR=vcfR, popmap = popmap,
                              thresholds = c(.75,.85,.95), clustering = FALSE)
#>1.43% of SNPs fell below a completeness cutoff of 0.75 and were removed from the VCF
#cutoff is specified, filtered vcfR object will be returned
#38.65% of SNPs fell below a completeness cutoff of 0.85 and were removed from the VCF
#cutoff is specified, filtered vcfR object will be returned
#85.85% of SNPs fell below a completeness cutoff of 0.95 and were removed from the VCF
#check what t-SNE clustering looks like at an 85% threshold
miss<-assess_missing_data_tsne(vcfR=vcfR, popmap = popmap,
                               thresholds = c(.75,.85,.95), clustering = FALSE)
#>cutoff is specified, filtered vcfR object will be returned
#1.43% of SNPs fell below a completeness cutoff of 0.75 and were removed from the VCF
#cutoff is specified, filtered vcfR object will be returned
#38.65% of SNPs fell below a completeness cutoff of 0.85 and were removed from the VCF
#cutoff is specified, filtered vcfR object will be returned
#85.85% of SNPs fell below a completeness cutoff of 0.95 and were removed from the VCF
#choose a cutoff resulting in an acceptable amount of missing data in each sample, and maximizes SNPs retained while minimizing overall missing data, and filter vcf
vcfR<-missing_by_snp(vcfR, cutoff = .75)
#for this, decide what you need! Jason is keeping 200 to evaluate between species diffeerentiation
#>cutoff is specified, filtered vcfR object will be returned
#38.65% of SNPs fell below a completeness cutoff of 0.85 and were removed from the VCF
#check how many SNPs and samples are left
vcfR
####minor allele and linkage filtering####
#investigate clustering patterns with and without a minor allele cutoff
#use min.mac() to investigate the effect of multiple cutoffs
vcfR.mac<-min_mac(vcfR = vcfR, min.mac = 1)
#> 32.61% of SNPs fell below a minor allele count of 2 and were removed from the VCF
#if we want to do differentiation and population structure, leave at 1?
#if you want to do things like selection, look into it?
#minor allele count
#assess clustering without MAC cutoff
#if you do a min.mac of 1 it doesn't matter, don't check
#if you do moore than 1 then you do want them
miss<-assess_missing_data_tsne(vcfR, popmap, clustering = FALSE)
#assess clustering with MAC cutoff
miss<-assess_missing_data_tsne(vcfR.mac, popmap, clustering = FALSE)

#t-SNE different clustering approach
###Finally, we will make sure that the depth and genotype quality look consistent across SNPs and samples, following our filtering pipeline.###
#plot depth per snp and per sample
dp <- extract.gt(vcfR, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)
#for that sample, how many of the SNPs are there. it's an index of missing data if it is white
#low missing data (purple) is good.
#plot genotype quality per snp and per sample
gq <- extract.gt(vcfR, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)
#yellow means high qualituy
####write out vcf files for downstream analyses####
#write out vcf with all SNPs
vcfR::write.vcf(vcfR, "aciurina.filtered.vcf.gz")
#linkage filter vcf to thin SNPs to one per 500bp
vcfR.thin<-distance_thin(vcfR, min.distance = 500)
#write out thinned vcf
vcfR::write.vcf(vcfR.thin, "aciurinafiltered.thinned.vcf.gz")

